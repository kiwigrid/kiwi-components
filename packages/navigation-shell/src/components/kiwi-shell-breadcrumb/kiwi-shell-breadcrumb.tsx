import { Component, h } from '@stencil/core';
import state from '../kiwi-navigation-shell/kiwi-navigation-shell.store';

@Component({
  tag: 'kiwi-shell-breadcrumb',
  shadow: false,
})
export class KiwiShellBreadcrumb {
  render(): JSX.Element | null {
    if (state.breadcrumb.length === 0) {
      return null;
    }

    return (
      <kiwi-breadcrumb>
        {state.breadcrumb.map((crumb) => {
          const { routeKey, data, label, labelOnly } = crumb ?? {};

          return (
            <kiwi-breadcrumb-item key={`kiwi-breadrumb-item-${routeKey}`}>
              {label !== undefined ? (
                <kiwi-shell-link
                  routeKey={routeKey}
                  routeData={data}
                  labelOnly={labelOnly}
                >
                  <span slot="kiwi-shell-link-content">{label}</span>
                </kiwi-shell-link>
              ) : (
                <kiwi-shell-link
                  routeKey={routeKey}
                  routeData={data}
                  labelOnly={labelOnly}
                />
              )}
            </kiwi-breadcrumb-item>
          );
        })}
      </kiwi-breadcrumb>
    );
  }
}
