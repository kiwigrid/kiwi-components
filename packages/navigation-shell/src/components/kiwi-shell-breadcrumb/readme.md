# kiwi-shell-breadcrumb

Renders a breadcrumb based on the current route history.

See the [Navigation Shell Example](../kiwi-navigation-shell/readme.md) for how to manipulate the breadcrumb.

<!-- Auto Generated Below -->


## Dependencies

### Depends on

- [kiwi-shell-link](../kiwi-shell-link)

### Graph
```mermaid
graph TD;
  kiwi-shell-breadcrumb --> kiwi-shell-link
  style kiwi-shell-breadcrumb fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
