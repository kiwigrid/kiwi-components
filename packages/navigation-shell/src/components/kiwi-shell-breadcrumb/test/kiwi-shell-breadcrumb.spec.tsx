import { newSpecPage } from '@stencil/core/testing';
import state, {
  dispose,
  init,
  makeLink,
  RouteHistory,
} from '../../kiwi-navigation-shell/kiwi-navigation-shell.store';
import { KiwiShellBreadcrumb } from '../kiwi-shell-breadcrumb';

describe('kiwi-shell-breadcrumb', () => {
  const initialBreadcrumb: RouteHistory = [
    { routeKey: 'deep' },
    { routeKey: 'down' },
  ];

  beforeEach(() => {
    init(
      [
        {
          routeKey: 'deep',
          label: 'Deep',
          url: '/deep',
          handler: () => [],
        },
        {
          routeKey: 'down',
          label: 'Down',
          url: '/deep/down',
          handler: () => [{ routeKey: 'deep' }],
        },
        {
          routeKey: 'rabbit-hole',
          label: 'The Rabbit Hole',
          url: '/deep/down/rabbit-hole',
          handler: () => [{ routeKey: 'deep' }, { routeKey: 'down' }],
        },
      ],
      initialBreadcrumb,
      'deep',
    );
  });

  afterEach(() => {
    dispose();
  });

  it('renders initialBreadcrumb', async () => {
    const crumb = await newSpecPage({
      components: [KiwiShellBreadcrumb],
      html: `<kiwi-shell-breadcrumb></kiwi-shell-breadcrumb>`,
    });

    expect(state.breadcrumb).toEqual(initialBreadcrumb);

    expect(crumb.root).toMatchSnapshot();
  });

  it('renders updated breadcrumb', async () => {
    const crumb = await newSpecPage({
      components: [KiwiShellBreadcrumb],
      html: `<kiwi-shell-breadcrumb></kiwi-shell-breadcrumb>`,
    });

    expect(state.breadcrumb).toEqual(initialBreadcrumb);

    const [, , onClick] = await makeLink('rabbit-hole', {});
    onClick(new MouseEvent('click'));

    await crumb.waitForChanges();

    expect(state.breadcrumb).toEqual([
      { routeKey: 'deep' },
      { routeKey: 'down' },
      {
        data: {},
        label: 'The Rabbit Hole',
        labelOnly: true,
        routeKey: 'rabbit-hole',
      },
    ]);

    expect(crumb.root).toMatchSnapshot();
  });
});
