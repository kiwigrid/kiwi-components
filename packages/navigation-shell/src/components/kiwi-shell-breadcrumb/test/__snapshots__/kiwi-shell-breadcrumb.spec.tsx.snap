// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`kiwi-shell-breadcrumb renders initialBreadcrumb 1`] = `
<kiwi-shell-breadcrumb>
  <kiwi-breadcrumb>
    <kiwi-breadcrumb-item>
      white
    </kiwi-breadcrumb-item>
    <kiwi-breadcrumb-item>
      rabbits
    </kiwi-breadcrumb-item>
    <kiwi-breadcrumb-item>
      burrow
    </kiwi-breadcrumb-item>
  </kiwi-breadcrumb>
</kiwi-shell-breadcrumb>
`;

exports[`kiwi-shell-breadcrumb renders updated breadcrumb 1`] = `
<kiwi-shell-breadcrumb>
  <kiwi-breadcrumb>
    <kiwi-breadcrumb-item>
      deep
    </kiwi-breadcrumb-item>
    <kiwi-breadcrumb-item>
      down
    </kiwi-breadcrumb-item>
    <kiwi-breadcrumb-item>
      the
    </kiwi-breadcrumb-item>
    <kiwi-breadcrumb-item>
      rabbit-hole
    </kiwi-breadcrumb-item>
  </kiwi-breadcrumb>
</kiwi-shell-breadcrumb>
`;
